Add withStream() in Streamlet to support stream selection

nwangtw opened the pull request:
Usage:

Assuming an operator creates two streams with ids "stream1" and "stream2" and it is managed by streamletWithTwoStreams (with type Streamlet<T>), it can be used like:

streamletWithTwoStreams
  .withStream("stream1")
  ......
  .log();

streamletWithTwoStreams
  .withStream("stream2")
  ......
  .log();

erenavsarogullari commented on heron/api/tests/java/org/apache/heron/streamlet/impl/StreamletImplTest.java:
```diff
@@ -81,6 +82,22 @@ public void testBasicParams() throws Exception {
   }
 
+  @Test
+  public void testWithStream() {
```
Test coverage can be useful for the following cases:
- if `streamId` is set duplicate (`test_stream2 == test_stream3`)
- if `streamId` is set `Bolt` level again as different from `Spout`'s `streamId`

erenavsarogullari commented on heron/api/src/java/org/apache/heron/streamlet/impl/streamlets/ConsumerStreamlet.java:
```diff
@@ -34,10 +34,12 @@
  */
 public class ConsumerStreamlet<R> extends StreamletImpl<R> {
   private StreamletImpl<R> parent;
+  private String parentStream;
```
may `parentStreamId` be more meaningful?

erenavsarogullari commented on heron/api/src/java/org/apache/heron/streamlet/Streamlet.java:
```diff
@@ -71,6 +71,22 @@
+   * @param streamId The specified stream id
+   * @return Returns back the Streamlet with changed stream id
+   */
+  Streamlet<R> withStream(String streamId);
```
I think `streamId` offers to be distinguished streams, right? 
Also, wondering is there UI impact (if user would like to see `streamId` or not)?

nwangtw replied:
Yeah. A component can generates multiple different streams (such as 3 integer streams: odd, even, all) and the next component needs to specify which one to accept data.

UI doesn't show this information currently and it could be a useful feature to add. And grouping strategy is another important information on the edges.

erenavsarogullari replied:
Exactly. Vertices show Streamlet name and Edges can show both streamId and grouping strategy.

erenavsarogullari replied:
As long-term, i think we can raise an UI issue to be addressed this (to be shown both **streamId** and **grouping strategy**).

nwangtw replied:
Yep.

erenavsarogullari replied:
I have just created for this: #3117

erenavsarogullari commented on heron/api/src/java/org/apache/heron/streamlet/impl/StreamletImpl.java:
```diff
@@ -205,11 +207,37 @@ public int getNumPartitions() {
+   */
+  @SuppressWarnings("HiddenField")
+  @Override
+  public Streamlet<R> withStream(String streamId) {
```
Also wondering the case: if `Bolt` points invalid `streamId` (not matched with `parentStreamId`) so could **validation** be required? (to cover this case either **fail-fast** or **warning message** due to redundant Bolt)?

nwangtw replied:
Agreed. Failfast is desirable for sure.

Validation is an option question for now. Currently all streamlet operators receive and emit tuples to default stream.

The stream ids are in run time information now. So we need to be careful about where the validation should be done and how to apply it.

nwangtw replied:
Heron has a check already and it runs during submission time. Not as good as compile time but should be fast enough. :)

erenavsarogullari commented on heron/api/src/java/org/apache/heron/streamlet/Streamlet.java:
```diff
@@ -71,6 +71,22 @@
+   * Gets the stream id of this Streamlet.
+   * @return the stream id of this Streamlet
+   */
+  String getStreamId();
```
Also IT coverage will be useful to verify end2end so will be addressing in the light of #3106

nwangtw replied:
In fact, I am thinking of adding the split() operator now with this PR so that we can add a proper IT for it. I prefer to break PR down to single functionality but in this case it could be useful.

erenavsarogullari replied:
Sure, IT coverage is too important due to aim end2end. I think it should be part of the code review checklist for incoming PRs ;) 

On the other hand, do you have template for `split` function signature?

nwangtw replied:
Did a quick search, I think it is pretty close to the split operator in Flink (https://ci.apache.org/projects/flink/flink-docs-release-1.6/dev/stream/operators/).

In fact, select() seems to be a better name than "withStream". :D I will think about it.

Neng gives a feedback about the withStream() function as a new user and he made a good point that we need to handle both cases:
1.
  stream.map();
  stream.withStream("a").map();
  stream.map();   // this should be the default stream
  stream.withStream("a").map();

2.
  steam2 = stream.withString("a");
  stream2.map();   // this should be the new stream
  stream.map();   // this should be the default stream

And this is not supported with the current implementation since the this object is muted. I am going to see what is the best way and update this PR.

erenavsarogullari replied:
Yep, i was having a look for Flink implementation as well. It supports child and parent Stream hierharchy to support `select` functions through `split` function.

I think more make sense `split` function naming convention and following can be as an alternative: 

```
Streamlet<Integer> numberStreamlets = ...
// user can apply his custom split logic
Map<String, Streamlet> splittedStreamlets = numberStreamlets.split(new SerializableSplitter())

splittedStreamlets.get("even")
splittedStreamlets.get("odd")
```

If we have proposal document by covering the all potential use-cases, that will be useful as well ;)

nwangtw replied:
"Map<String, Streamlet<Integer>>"  won't really work because right now one Streamlet is mapped to one operator which is a Bolt component. In our case, there is only one component (hence one streamlet object) with multiple streams.

I had to add a StreamletShadow class to create companion objects of the main Streamlet object in my implementation.

nwangtw replied:
In fact, still no....

The problem is the .setName and .setPartition functions, like:

numberStreamlets.split(splitter).setName("split").setPartition(10);

the split operator is one single bolt and all entries in Map<String, Streamlet> points to the same bolt. It might be ok to set name and partition like:
Map<String, Streamlet> splittedStreamlets  = numberStreamlets.split(splitter);

splittedStreamlets.get("even").setName("split")
splittedStreamlets.get("even").setPartition(partition)

but it looks really bad this way.

erenavsarogullari commented on heron/api/src/java/org/apache/heron/streamlet/impl/operators/SplitOperator.java:
```diff
@@ -0,0 +1,67 @@
+  public SplitOperator(Map<String, SerializablePredicate<R>> splitFns) {
+    this.splitFns = splitFns;
+    // Make sure stream ids are not empty
+    for (String stream: splitFns.keySet()) {
```
This is already checked upper level

erenavsarogullari commented:
Thanks again @nwangtw for all your effort.
Overall looks ok for me and also, split function logic looks as common pattern with Flink.

nwangtw commented:
Resolved comments, except the UT for wrong stream id. And I am going to add a UT for the new StrealetShadow class.

nwangtw commented:
- Added stream id check in withStream() function
- Added unit test for calling withStream() with invalid (unavailable) stream id.
- Add unit test for StreamletShadow.
